[Green CI] Fix crash in DistributedAsyncInsert when connection is empty

pamarcos opened the pull request:
Fix crash in DistributedAsyncInsert when connection is empty

### Changelog category (leave one):
- Critical Bug Fix (crash, LOGICAL_ERROR, data loss, RBAC)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix crash in DistributedAsyncInsert when connection is empty

Closes https://github.com/ClickHouse/ClickHouse/issues/66740

pamarcos committed:
Fix crash when the connection is empty

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #67219 from pamarcos/ClickHouse:fix-crash-in-distributed-async-insert-directory-queue -->
*This is an automated comment for commit 42384af0ef38dd326337e8cf18327871924f7359 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67219/42384af0ef38dd326337e8cf18327871924f7359/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/9ff1492245ee1805c6062f57ffb243106a839399/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Fix style check

pamarcos commented:
The [only test that failed](https://s3.amazonaws.com/clickhouse-test-reports/67219/981135bfb104b5ecfa0f1da5533e3d12f6850838/stateless_tests__msan__[2_4].html) is already [reported](https://github.com/ClickHouse/ClickHouse/issues/67169) and [fixed](https://github.com/ClickHouse/ClickHouse/pull/67173)

divanik commented on src/Storages/Distributed/DistributedAsyncInsertBatch.cpp:
```diff
@@ -245,6 +246,9 @@ void DistributedAsyncInsertBatch::sendBatch(const SettingsChanges & settings_cha
                 connection = std::move(result.front().entry);
                 compression_expected = connection->getCompression() == Protocol::Compression::Enable;
 
+                if (connection.isNull())
```
It really seems as a logical error and ideally should be debugged. As far as I see invariant that should be satisfied is !r.entry.isNull() && r.is_usable && !(settings.distributed_insert_skip_read_only_replicas && r.is_readonly). Could you double check this and change condition to this one, please?

pamarcos replied:
Done! You're right. Because of [this line](https://github.com/ClickHouse/ClickHouse/blob/1480275ade1919651877e087fee3fda19924bc9f/src/Common/PoolWithFailoverBase.h#L303), that's the invariant that should be satisfied moving forward

divanik reviewed (commented):
1. Also, please, add LogicalError to https://github.com/ClickHouse/ClickHouse/blob/a05312c6541d4c761fd5c242dba372d2cff5e842/src/Common/PoolWithFailoverBase.h#L324-L325 checking that up_to_date_count <= try_results.size(). Till now I haven't found any places where the invariant can be broken.
2. Could you explain to me briefly, how you found out that seg fault in the issue is connected with the LogicalError you added in this PR?

pamarcos commented:
> Could you explain to me briefly, how you found out that seg fault in the issue is connected with the LogicalError you added in this PR?

Sure. I haven't been able to reproduce the issue, but according to the stack trace, the NULL pointer access happens in `void DistributedAsyncInsertDirectoryQueue::processFile`. Analyzing the source code in that function looking for pointer accesses, the only one that makes sense is the connection, because the `pool` is moved in its constructor and I didn't see in the logic how that could ever be NULL.

pamarcos committed:
Ensure there's no logical error using PoolWithFailoverBase::getMany

Prevents a segfault in case the connection is null, but also checks
all other conditions to ensure the logic is the expected one.

pamarcos commented:
> 1. Also, please, add LogicalError to https://github.com/ClickHouse/ClickHouse/blob/a05312c6541d4c761fd5c242dba372d2cff5e842/src/Common/PoolWithFailoverBase.h#L324-L325

Done in https://github.com/ClickHouse/ClickHouse/commit/eaede913b2ae0d5a4cb02ae82db20b6011270fe6#diff-f206c93448029d3b6e8aa3baa8834d0909740bd20709f984e1e06cfaecc0f6cdR330

kssenii commented:
@pamarcos  Could this PR have caused this error ` Logical error: 'Received from kng4alm55c.us-east-2.aws.clickhouse-staging.com:9440. DB::Exception: Unexpected error for async deduplicated insert, please check error logs.` ? I saw this error today for the first time, no similar issue found. https://s3.amazonaws.com/clickhouse-test-reports/67361/2b79da36c0701bb9ca392fddd9129a7e0e04ef3f/stateless_tests__azure__asan__[2_3].html

pamarcos commented:
> @pamarcos Could this PR have caused this error ` Logical error: 'Received from kng4alm55c.us-east-2.aws.clickhouse-staging.com:9440. DB::Exception: Unexpected error for async deduplicated insert, please check error logs.` ? I saw this error today for the first time, no similar issue found. [https://s3.amazonaws.com/clickhouse-test-reports/67361/2b79da36c0701bb9ca392fddd9129a7e0e04ef3f/stateless_tests__azure__asan__[2_3].html](https://s3.amazonaws.com/clickhouse-test-reports/67361/2b79da36c0701bb9ca392fddd9129a7e0e04ef3f/stateless_tests__azure__asan__%5B2_3%5D.html)

Hum. These changes do not include that trace, that's for sure. I downloaded the logs and checked that neither of the logical errors added by this PR happened. There's no

>Could not find enough connections for up-to-date results

and no

>Got an invalid connection result

There's another interesting error right before the one you shared:

```
2024.07.31 18:24:11.995609 [ 1460 ] {} <Error> system.query_log_sender.DistributedInsertQueue.default: Code: 241. DB::Exception: Received from kng4alm55c.us-east-2.aws.clickhouse-staging.com:9440. DB::Exception: Memory limit (for user) exceeded: would use 1.79 TiB (attempt to allocate chunk of 1513063 bytes), maximum: 20.60 GiB. OvercommitTracker decision: Memory overcommit has freed not enough memory..
```
